docs: Simplify process to add url redirects when renaming docs.

laurynmm opened the pull request:
Creates ~`zerver.lib.documentation_redirects.py`~ `zerver.lib.url_redirects.py` to record old and new URLs for help center, API, and policy documentation pages, as well as website pages, that have been renamed and need redirects.

This file is then used by `zproject.urls.py` to redirect links and ~by `zerver.lib.test_helpers.py` to exempt these old URLs from documentation tests~ by `zerver.tests.test_urls.py` to test that the redirects return successful responses.

Fixes #21946. Fixes #17897.

---

**Testing**:
- Manually reviewed all old URLs to be sure the redirects still work.
- Commenting out the redirects in the new file also causes current documentation tests to fail as `/help/add-custom-emoji` is still linked to in `docs.subsystems.markdown.md` and `templates.zerver.for-open-source.html` (*Note: could fix this in follow-up PR or as a separate commit here*).


**To do**:
- [x] Add "Moving an existing article" (or similar sub-section title to match "Updating an existing article" and "Adding a new article") to [guide to writing help center documentation](https://zulip.readthedocs.io/en/latest/documentation/helpcenter.html#guide-to-writing-help-center-articles).
- [x] Possibly add some documentation within `documentation_redirects.py` for contributors when updating the file for the first time.

---

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

timabbott commented on zerver/lib/test_helpers.py:
```diff
@@ -518,6 +502,8 @@ def find_pattern(pattern: Any, prefixes: List[str]) -> None:
             "scim/v2/Groups/.search",
             *(webhook.url for webhook in WEBHOOK_INTEGRATIONS if not include_webhooks),
+            # We exempt help center and api docs that have been renamed / redirected
+            *(redirect.old_url for redirect in DOCUMENTATION_REDIRECTS),
```
I think there's also some number that we have actual tests for in `test_docs` or `test_urls`; can't remember which. 

I think it's possible we could just replace this with changing that code to just test them all -- testing a redirect should be very cheap.

laurynmm replied:
I could find no tests for any of these `redirect.old_urls` in the codebase. Should I write one that loops through the list?

laurynmm replied:
I found a test in `test_docs.py` for policy page redirects, so I used that as a template and wrote one that loops through the `DOCUMENTATION_REDIRECTS` for a success response and a common header in both the api and help center documentation ("Zulip homepage").

timabbott replied:
Sounds good!

timabbott commented:
This looks great; posted one comment, and then I think we should mention this feature somewhere in the "Writing help center documentation" docs, so it's discoverable.

laurynmm commented:
**Update**:
- Added test for these redirects in `test_docs.py`, which eliminates needing to add them to the exempt patterns list in `test_helpers.py`.
- Added [a new section](https://zulip--22444.org.readthedocs.build/en/22444/documentation/helpcenter.html#redirecting-an-existing-article) to the "Writing help center documentation" docs.
- Added a comment in `documentation_redirects.py` that gives some basic context / information.

laurynmm commented:
@drrosa - Would you be up for doing a review of this PR since Alya's out for a few more days? Ideally, thinking of the new system from the point of view of contributors who would be focused mostly on documentation updates/work (see descriptions of two linked issues).

drrosa commented:
Yeah, I'm up for it!

timabbott commented on zerver/lib/test_helpers.py:
```diff
@@ -482,24 +482,7 @@ def find_pattern(pattern: Any, prefixes: List[str]) -> None:
-            "help/test-mobile-notifications",
-            "help/troubleshooting-desktop-notifications",
-            "help/web-public-streams",
             "for/working-groups-and-communities/",
```
This one is a redirect too; it just was in a different place in the file.

laurynmm replied:
Added in updated version, which extends this new system to include all redirects that were in `zproject.urls.py`.

laurynmm replied:
**Update**: Moved to 2nd commit because this is a corporate landing page URL.

timabbott commented on zerver/lib/documentation_redirects.py:
```diff
@@ -0,0 +1,55 @@
+    DocRedirect("help/old", "/help/new"),
+
+    Note that the old URL should not start with a forward slash,
+    while the new URL should.
```
It could be helpful to couple checks like `not assert old_url.startswith("/")` inside the loop in `zproject/urls.py` to catch this class of error.

laurynmm replied:
Decided to go with the latter option, so the forward slash is stripped from the string in `zproject.urls.py`, which makes the format for the old and new URLs the same.

timabbott commented on docs/documentation/helpcenter.md:
```diff
@@ -148,6 +152,32 @@ with modals, etc. should never go in the help center documentation.
+From time to time, there is a reason to rename articles in the help center or
+REST API documentation. This breaks incoming links, including from other
+branches of the repository that haven't been rebased, and more importantly
+from previous versions of Zulip.
```
We should mention the Zulip blog as a possible source as well.

laurynmm replied:
I believe you meant that published blog posts would be a source for broken links and not something we would rename/move. I updated the text to include that information.

timabbott commented on docs/documentation/helpcenter.md:
```diff
@@ -148,6 +152,32 @@ with modals, etc. should never go in the help center documentation.
+To make it easier for contributors make these changes, there is only one place
+where you need to add a redirect: `zerver/lib/documentation_redirects.py`.
+
+Once you've renamed the file (e.g. `foo.md` is now `bar.md`) in your branch,
```
We could spell this `git mv path/to/foo.md path/to/bar.md`? Not sure if it's helpful.

laurynmm replied:
Updated for that suggestion.

timabbott commented:
I posted a few quick comments; otherwise lgtm; please add the "integration review" label to this once others are done reviewing so I know it's ready for me to merge.

drrosa reviewed (changes requested):
I've suggested a few tweaks to the documentation. Everything else looks good to me!

laurynmm commented on docs/documentation/helpcenter.md:
```diff
@@ -112,6 +112,10 @@ existing help center articles:
 
+If your updates to the existing article will change the name of the
+markdown file, then see section below on [redirecting an existing
+article](#redirecting-an-existing-article)
```
Fixed.

laurynmm commented on docs/documentation/helpcenter.md:
```diff
@@ -148,6 +152,32 @@ with modals, etc. should never go in the help center documentation.
+From time to time, there is a reason to rename articles in the help center or
+REST API documentation. This breaks incoming links, including from other
+branches of the repository that haven't been rebased, and more importantly
+from previous versions of Zulip.
```
Thanks for these suggestions. I think Tim meant that the blog would be a source for broken links and not something that we would rename/redirect. Also, since there are multiple repositories (zulip-mobile, zulip-desktop, etc.), I thought your rewording was a bit confusing. The updated version is now ...

> From time to time, we might want to rename an article in the help
center, or REST API documentation. This change will break incoming
links, including links in published Zulip blog posts, links in other
branches of the repository that haven't been rebased, and more
importantly links from previous versions of Zulip.

laurynmm commented on docs/documentation/helpcenter.md:
```diff
@@ -148,6 +152,32 @@ with modals, etc. should never go in the help center documentation.
+from previous versions of Zulip.
+
+To make it easier for contributors make these changes, there is only one place
+where you need to add a redirect: `zerver/lib/documentation_redirects.py`.
```
Yes, this is great. I simplified a little bit more and ended up with:

> To fix these broken links, you can easily add a URL redirect in:
`zerver/lib/url_redirects.py`.

laurynmm commented on docs/documentation/helpcenter.md:
```diff
@@ -148,6 +152,32 @@ with modals, etc. should never go in the help center documentation.
+Once you've renamed the file (e.g. `foo.md` is now `bar.md`) in your branch,
+add a new `DocRedirect` to the list with the old URL (`"help/foo"`) and the
+new URL (`"/help/bar"`). Note that the old URL should not start with a forward
+slash, but the new URL should. Both should be in quotation marks.
```
I ended up revising this section a bunch and adding some more information as well. See screenshot in general update comment.

laurynmm commented:
@drrosa - Thanks for the review and thoughtful suggestions!

**Updates**:
- I decided to extend this new system to work and test for all the `RedirectView` instances in `zproject.urls.py`. So it now is not only for documentation pages, but also website pages. This prompted some renaming and restructuring from the previous version of the PR (e.g., `documentation_redirects.py` is now `url_redirects.py`).
- I extended the tutorial section a bit to include information about redirecting to sections of other articles and still replacing references to the old URL in the repository. And I updated the tutorial for the name changes mentioned in the previous bullet point.

@alya - I think this is ready for you to review and try out when you have a chance.

<details>
<summary>Updated tutorial addition to "Writing help center documentation</summary>

![Screenshot from 2022-08-19 16-46-26](https://user-images.githubusercontent.com/63245456/185644959-7c1fe698-d4d2-4571-9af1-2c3c29d7d873.png)
</details>

drrosa commented:
@laurynmm The updates look good to me. Thanks for pinging me for review. I think it's a nice system and looks like it's going to be very useful!

laurynmm commented:
**Update**: Since #22587 was merged and it moved the URL redirects for landing pages to `/corporate/urls.py`, I changed this PR to have two commits.

The first one sets up the URL redirect system for documentation page redirects, which are all in `/zproject/urls.py`. The second one extends that system to the landing page URL redirects, which are all in `/corporate/urls.py`.

I had originally updated this PR to include those landing page redirects since there was one in `/zerver/lib/test_helpers.py` that was included in the list of help center documentation redirects (see [this comment](https://github.com/zulip/zulip/pull/22444#discussion_r947336347)).

drrosa commented:
> **Update**: Since #22587 was merged and it moved the URL redirects for landing pages to `/corporate/urls.py`, I changed this PR to have two commits.
> 
> The first one sets up the URL redirect system for documentation page redirects, which are all in `/zproject/urls.py`. The second one extends that system to the landing page URL redirects, which are all in `/corporate/urls.py`.

The update looks good to me!

> I posted a few quick comments; otherwise lgtm; please add the "integration review" label to this once others are done reviewing so I know it's ready for me to merge.

@timabbott I think this PR is ready for you.
@zulipbot add "integration review"

timabbott commented:
This is awesome, merged, thanks @laurynmm!

drrosa commented:
Yeah, I'm up for it!
